allennlp/models/vilbert.py

```diff
@@ -0,0 +1,836 @@
+        visual_features: torch.Tensor,
+        box_coordinates: torch.Tensor,
+        image_id: List[List[str]],
+        identifier: List[str],
+        sentence_field: TextFieldTensors,
+        denotation: torch.Tensor = None,
+    ) -> Dict[str, torch.Tensor]:
+
+        batch_size, num_images, num_objects, feature_size = visual_features.size()
+        assert num_objects == 36 and feature_size == 2048
+        assert num_images == 2
+
+        input_ids = sentence_field["tokens"]["token_ids"]
+        token_type_ids = sentence_field["tokens"]["type_ids"]
+        attention_mask = sentence_field["tokens"]["mask"]
+        # We will always have an equal number of images and boxes, so no masking necessary.
+        image_attention_mask = torch.ones_like(box_coordinates)[:, :, :, 0]
+
+        # (batch_size, num_tokens, embedding_dim)
+        embedding_output = self.embeddings(input_ids, token_type_ids)
+        num_tokens = embedding_output.size(1)
+
+        # Repeat the embedding dimension, so that the TimeDistributed works out ok
+        embedding_output = embedding_output.unsqueeze(1).expand(-1, 2, -1, -1)
+        attention_mask = attention_mask.unsqueeze(1).expand(-1, 2, -1)
+
+        # We create a 3D attention mask from a 2D tensor mask.
+        # Sizes are [batch_size, 1, 1, to_seq_length]
+        # So we can broadcast to [batch_size, num_heads, from_seq_length, to_seq_length]
+        # this attention mask is more simple than the triangular masking of
+        # causal attention used in OpenAI GPT, we just need to prepare the
+        # broadcast dimension here.
+        extended_attention_mask = attention_mask.unsqueeze(2).unsqueeze(3)
+        extended_image_attention_mask = image_attention_mask.unsqueeze(2).unsqueeze(3)
+
+        extended_attention_mask2 = attention_mask.unsqueeze(3)
+
+        # Since attention_mask is 1.0 for positions we want to attend and 0.0 for
+        # masked positions, this operation will create a tensor which is 0.0 for
+        # positions we want to attend and -10000.0 for masked positions.
+        # Since we are adding it to the raw scores before the softmax, this is
+        # effectively the same as removing these entirely.
+        extended_attention_mask = extended_attention_mask.to(
+            dtype=next(self.parameters()).dtype
+        )  # fp16 compatibility
+        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
+
+        extended_attention_mask2 = extended_attention_mask2.to(
+            dtype=next(self.parameters()).dtype
+        )  # fp16 compatibility
+
+        extended_image_attention_mask = extended_image_attention_mask.to(
+            dtype=next(self.parameters()).dtype
+        )  # fp16 compatibility
+        extended_image_attention_mask = (1.0 - extended_image_attention_mask) * -10000.0
+
+        co_attention_mask = torch.zeros(batch_size, num_images, feature_size, num_tokens).type_as(
+            extended_image_attention_mask
+        )
```

Comment:
```suggestion
        co_attention_mask = torch.zeros(batch_size, num_images, feature_size, num_tokens, dtype=extended_image_attention_mask.dtype)
```
This avoids another temporary tensor.

Reply:
Done.